modules/client-common/src/main/java/org/apache/ignite/client/ClientMessageDecoder.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.client;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.CharsetUtil;
+import org.apache.ignite.lang.IgniteException;
+import org.msgpack.core.MessageFormat;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Decodes full client messages:
+ * 1. MAGIC for first message.
+ * 2. Payload length (varint).
+ * 3. Payload (bytes).
+ */
+public class ClientMessageDecoder extends ByteToMessageDecoder {
+    /** Magic bytes before handshake. */
+    public static final byte[] MAGIC_BYTES = new byte[]{0x49, 0x47, 0x4E, 0x49}; // IGNI
+
+    /** Data buffer. */
+    private byte[] data = new byte[4]; // TODO: Pooled buffers IGNITE-15162.
+
+    /** Remaining byte count. */
+    private int cnt = -4;
+
+    /** Message size. */
+    private int msgSize = -1;
+
+    /** Magic decoded flag. */
+    private boolean magicDecoded;
+
+    /** Magic decoding failed flag. */
+    private boolean magicFailed;
+
+    /** Message size varint format. */
+    private MessageFormat sizeFormat = null;
+
+    /** First byte of the message. */
+    private byte firstByte;
+
+    /** {@inheritDoc} */
+    @Override protected void decode(ChannelHandlerContext ctx, ByteBuf byteBuf, List<Object> list) {
+        if (!readMagic(byteBuf))
+            return;
+
+        while (read(byteBuf))
+            list.add(ByteBuffer.wrap(data));
```

Comment:
It is strange to wrap onheap data to ByteBuf. This is uneffective. Why don't pass ByteBuf as is?

Reply:
We need to find the most effective way to bridge MsgPack and Netty while employing buffer pooling. I'll investigate this as part of IGNITE-15162. Thanks for all the suggestions, I'm keeping notes. Let's keep things as is here to reduce scope.